design/0001-frr.md

```diff
@@ -0,0 +1,347 @@
+# MLB-0001: BGP FRR enablement
+
+## Summary
+
+The purpose of this enhancement is to use [Free Range
+Routing](https://frrouting.org/) (FRR) to provide BGP functionality in MetalLB.
+When directed to, MetalLB will publish prefixes via FRR rather than
+MetalLB’s current built-in BGP speaker.
+
+## Motivation
+
+The motivation for this enhancement is to provide an alternative
+production-ready BGP agent for use in MetalLB. Overall, this should reduce the
+maintenance cost for the MetalLB project by reducing the effort required for
+adding additional BGP features and bug fixes. For example, there are a number of
+issues in the current backlog that may be addressed by using FRR:
+
+* [Add support for MP BGP encoding for IPv4 and IPv6](https://github.com/metallb/metallb/pull/590)
+* [BFD support](https://github.com/metallb/metallb/issues/396)
+* [BGP Failover too slow](https://github.com/metallb/metallb/issues/485)
+
+FRR is a mature Linux Foundation routing protocol suite based on Quagga that has
+been used in many production deployments. As such, it has been proven in terms
+of its maturity, flexibility (as can be seen by the [broad range of features](http://docs.frrouting.org/en/latest/bgp.html)
+it supports), scalability, security, reliability and performance. It also
+provides detailed logging features to aid debugging.
+
+### Goals
+
+* Provide a configuration option to allow use of FRR as a BGP agent in a MetalLB
+  deployment at cluster installation time.
+* Ensure feature parity with current BGP implementation.
+* Reinstate thorough BGP testing infrastructure.
+
+### Non-Goals
+
+* Coexistence of two types of BGP agents (e.g. MetalLB native BGP agent AND FRR
+  BGP agent) is not a goal of this work
+* The purpose of this is not to provide an interface to plug any BGP agent into
+  MetalLB. Although, this may provide a useful template for others to do this.
+* This will not support runtime changing of BGP agent type.
+
```

Comment:
Non-Goal?  * Allowing anything else beside MetalLB to use FRR?

Or is it a goal (in MetalLB repo??) to let something else e.g. HAProxy/nginx/whatever inform FRR about prefix(es) to advertise as well?

Reply:
Yeah, this is a good addition and msherif1234 mentioned something similar above. I think it would be worth adding it as a goal and spending some time ensuring that this can be done. Thanks.

Reply:
Could you then run FRR w/out MetalLB?   Just use e.g. HAProxy or nothing... just advertise prefixes?

Reply:
You could, but you still need something that understands what to advertise, and where to advertise it from.  That's the layer that MetalLB solves for the LoadBalancer Service use case.

Reply:
@mccv1r0 @russellb I have been thinking about this some more. I think it might be better to not allow other things to advertise prefixes - at least by default - as this will open the door to unusual configurations that may introduce weird bugs that will be difficult for the upstream community to support and debug. It could potentially make an unfriendly "out-of-the-box experience" if a user attempted to do this. As such I think this should be a _non-goal_.

However, I think this is an important use case and may be a common deployment model. Therefore, we need to consider this during development in order to not preclude the ability for a _smart_ consumer of MetalLB to be able to easily engineer this if required. It could also be that, with time, we do allow this capability. What do you think?